src/persist/src/s3.rs

```diff
@@ -163,30 +163,216 @@ impl S3BlobCore {
             Ok(object) => object,
             Err(SdkError::ServiceError { err, .. }) if err.is_no_such_key() => return Ok(None),
-            Err(err) => return Err(Error::from(err.to_string())),
+            Err(err) => return Err(Error::from(format!("s3 get meta err: {}", err))),
+        };
+        let num_parts = if first_part.parts_count() == 0 {
+            // For a non-multipart upload, parts_count will be 0. The rest of
+            // the code works perfectly well if we just pretend this was a
+            // multipart upload of 1 part.
+            1
+        } else {
+            first_part.parts_count()
         };
+        tracing::trace!(
+            "s3 download first header took {:?} ({} parts)",
+            start_overall.elapsed(),
+            num_parts
+        );
+        if num_parts < 1 {
+            return Err(format!("unexpected number of s3 object parts: {}", num_parts).into());
+        }
 
-        let val = object
-            .body
-            .collect()
-            .await
-            .map_err(|e| format!("collecting byte buffer: {}", e))?
-            // TODO: `into_bytes().to_vec()` results in one extra copy than
-            // necessary. Changing `Blob::get` to return `Bytes` directly would
-            // be more efficient, or adding a `into_vec()` method upstream.
-            .into_bytes()
-            .to_vec();
+        // TODO: Plumb an Arc<AsyncRuntime> on S3BlobCore instead. I tried but
+        // was getting a "Cannot drop a runtime in a context where blocking is
+        // not allowed" error that I didn't want to get distracted with. All the
+        // calls into the s3 library require that this context is set anyway, so
+        // I suppose this is no worse than where we started.
+        let async_runtime = AsyncHandle::try_current().map_err(|err| err.to_string())?;
+
+        // Continue to fetch the first part's body while we fetch the other
+        // parts.
+        let start_part_body = Instant::now();
+        let mut part_bodies = Vec::new();
+        let first_body_len = usize::try_from(first_part.content_length)
+            .map_err(|err| format!("unexpected s3 content_length: {}", err))?;
+        let mut total_body_len = first_body_len;
+        part_bodies.push(
+            async_runtime.spawn_named(
+                || "persist_s3blob_get_body",
+                first_part
+                    .body
+                    .collect()
+                    .map(move |res| (start_part_body.elapsed(), first_body_len, res)),
+            ),
+        );
+
+        if num_parts > 1 {
+            // Fetch the headers of the rest of the parts. (Starting at part 2
+            // because we already did part 1.)
+            let start_headers = Instant::now();
+            let mut part_futs = Vec::new();
+            for part_num in 2..=num_parts {
+                let part_fut = async_runtime.spawn_named(
+                    || "persist_s3blob_get_header",
```

Comment:
I just want to double check my understanding -- this future spawns a request that will return the header and the body right? It will just return the header first, and then we kick off another future to request to get the body? Concretely: its not like we spawn 2 * n_parts requests one for each header and one for each body?